alf/batch_norm.py

```diff
@@ -0,0 +1,330 @@
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import torch
+import torch.nn as nn
+import torch.nn.functional as F
+from typing import List, Union
+import types
+
+from alf.utils.common import warning_once
+
+
+class _NormBase(nn.Module):
+    """Base of BatchNorm supporting RNN."""
+
+    def __init__(self,
+                 num_features: int,
+                 eps: float = 1e-5,
+                 momentum: float = 0.1,
+                 affine: bool = True,
+                 track_running_stats: bool = True):
+        super().__init__()
+        self._num_features = num_features
+        self._eps = eps
+        self._momentum = momentum
+        self._affine = affine
+        self._track_running_stats = track_running_stats
+        if affine:
+            self._weight = nn.Parameter(torch.Tensor(num_features))
+            self._bias = nn.Parameter(torch.Tensor(num_features))
+        self._running_means = []
+        self._running_vars = []
+        self._num_batches_tracked = []
+        self.set_max_steps(1)
+        self._current_step = 0
+        self.reset_parameters()
+        self._clamped = False
+
+    def set_max_steps(self, max_steps: int):
+        """Set max steps to keeping running statistics.
+
+        Args:
+            max_steps: the maximum steps for which the batch norm running statistics
+                are maintained.
+        """
+        self._max_steps = max_steps
+        if not self._track_running_stats:
+            return
+        for i in range(len(self._running_means), max_steps):
+            self._running_means.append(torch.zeros(self._num_features))
+            self.register_buffer('_running_means%s' % i,
+                                 self._running_means[i])
+            self._running_vars.append(torch.ones(self._num_features))
+            self.register_buffer('_running_vars%s' % i, self._running_vars[i])
+            self._num_batches_tracked.append(torch.zeros((), dtype=torch.long))
+            self.register_buffer('_num_batches_tracked%s' % i,
+                                 self._num_batches_tracked[i])
+
+    def set_current_step(self, current_step: Union[torch.Tensor, int]):
+        """Use and/or update the running statistics at current_step for normalization.
+
+        Args:
+            current_step: the current step. If it is a Tensor, it means that
+                the current step for each sample in a batch.
```

Comment:
Slightly more specific about the shape of the input `current_step`? Also the constraint on "all steps must be the same for training" can be mentioned here.

Reply:
added